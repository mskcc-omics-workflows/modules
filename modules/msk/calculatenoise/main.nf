process CALCULATENOISE {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://ghcr.io/mikefeixu/sequence_qc:0.2.6' :
        'ghcr.io/mikefeixu/sequence_qc:0.2.6' }"

    input:
    tuple val(meta) , path(bam), path(bai), path(target_intervals)
    path(fasta)
    path(fai)

    output:
    tuple val(meta), path("${prefix}_acgt.tsv"),            emit: acgt
    tuple val(meta), path("${prefix}_by_substitution.tsv"), emit: substitution
    tuple val(meta), path("${prefix}_by_tlen.tsv"),         emit: bytlen
    tuple val(meta), path("${prefix}_del.tsv"),             emit: del
    tuple val(meta), path("${prefix}_n.tsv"),               emit: count
    tuple val(meta), path("${prefix}_positions.tsv"),       emit: positions
    tuple val(meta), path("${prefix}.html"),                emit: html
    tuple val(meta), path("${meta.id}_pileup.tsv"),         emit: pileup
    path "versions.yml",                                    emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}_noise"
    def reference_arg = fasta ? "--ref_fasta ${fasta}" : ''

    // Memory setting (optional)
    def avail_mem = 3000  // default in MB
    if (!task.memory) {
        log.info '[calculate_noise] Available memory not known - defaulting to 3GB. Specify process.memory to change this.'
    } else {
        avail_mem = (task.memory.mega * 0.8).intValue()
    }

    """
    sed -i 's/\r\$//' $target_intervals
    calculate_noise \\
        ${reference_arg} \\
        --bam_file $bam \\
        --bed_file $target_intervals \\
        --sample_id ${meta.id} \\
        --threshold 0.01 \\
        --truncate 1 \\
        --min_mapq 10 \\
        --min_basq 10 \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        calculatenoise: \$(calculate_noise --help 2>/dev/null || echo "Usage")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}_noise"
    """
    touch ${prefix}_acgt.tsv
    touch ${prefix}_by_substitution.tsv
    touch ${prefix}_by_tlen.tsv
    touch ${prefix}_del.tsv
    touch ${prefix}_n.tsv
    touch ${prefix}_positions.tsv
    touch ${prefix}.html
    touch ${meta.id}_pileup.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        calculatenoise: \$(calculate_noise --help 2>/dev/null || echo "Usage")
    END_VERSIONS
    """
}
