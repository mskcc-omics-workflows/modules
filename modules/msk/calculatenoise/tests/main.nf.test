nextflow_process {

    name "Test Process CALCULATENOISE"
    script "../main.nf"
    process "CALCULATENOISE"

    tag "modules"
    tag "modules_msk"
    tag "calculatenoise"

    test("chr22 - bam") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.test_data_mskcc['calculate_noise']['test_chr22_bam'], checkIfExists: true),
                    file(params.test_data_mskcc['calculate_noise']['test_chr22_bam_bai'], checkIfExists: true),
                    file(params.test_data_mskcc['calculate_noise']['test_bed'], checkIfExists: true),
                    ]
                input[1] = file(params.test_data_mskcc['calculate_noise']['test_chr22_fa'], checkIfExists: true)
                input[2] = file(params.test_data_mskcc['calculate_noise']['test_chr22_fa_fai'], checkIfExists: true)
                """
            }
        }

        then {
            def size = path(process.out.acgt[0][1]).size()
            def lines = path(process.out.acgt[0][1]).readLines()[0..1]
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.acgt[0][1]).name,
                    size,
                    lines,
                    process.out.versions,
                    file(process.out.acgt[0][1]).name,
                    file(process.out.substitution[0][1]).name,
                    file(process.out.bytlen[0][1]).name,
                    file(process.out.del[0][1]).name,
                    file(process.out.count[0][1]).name,
                    file(process.out.positions[0][1]).name,
                    file(process.out.html[0][1]).name,
                    file(process.out.pileup[0][1]).name
                    ).match()
                }
            )
        }
    }

    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used but keep the " - stub" suffix.
    test("chr22 - bam - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.test_data_mskcc['calculate_noise']['test_chr22_bam'], checkIfExists: true),
                    file(params.test_data_mskcc['calculate_noise']['test_chr22_bam_bai'], checkIfExists: true),
                    file(params.test_data_mskcc['calculate_noise']['test_bed'], checkIfExists: true),
                    ]
                input[1] = file(params.test_data_mskcc['calculate_noise']['test_chr22_fa'], checkIfExists: true)
                input[2] = file(params.test_data_mskcc['calculate_noise']['test_chr22_fa_fai'], checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
