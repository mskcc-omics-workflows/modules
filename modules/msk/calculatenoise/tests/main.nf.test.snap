{
    "chr22 - bam": {
        "content": [
            "test_noise_acgt.tsv",
            125,
            [
                "sample_id\tminor_allele_count\tmajor_allele_count\tnoise_fraction\tcontributing_sites",
                "test\t224\t2947046\t7.600253794189197e-05\t190"
            ],
            [
                "versions.yml:md5,51f2739b156b87202dedf84f6a82c787"
            ],
            "test_noise_acgt.tsv",
            "test_noise_by_substitution.tsv",
            "test_noise_by_tlen.tsv",
            "test_noise_del.tsv",
            "test_noise_n.tsv",
            "test_noise_positions.tsv",
            "test_noise.html",
            "test_pileup.tsv"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-09T23:47:29.902284"
    },
    "chr22 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_acgt.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_by_substitution.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_by_tlen.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_del.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_n.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_positions.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_pileup.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "8": [
                    "versions.yml:md5,1bb9b184fc77f055e623f2c5f3bc15c7"
                ],
                "acgt": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_acgt.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bytlen": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_by_tlen.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "count": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_n.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "del": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_del.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "html": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pileup": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_pileup.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "positions": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_positions.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "substitution": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_noise_by_substitution.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1bb9b184fc77f055e623f2c5f3bc15c7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-09T23:47:45.119775"
    }
}